File: benchmark/binary/x86-64/window/small/source/golang/function_pointers_and_strings.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

// Comparison function type
type CompareFunc func(int, int) int

// Compare functions
func compareAscending(a, b int) int {
	if a < b {
		return -1
	}
	if a > b {
		return 1
	}
	return 0
}

func compareDescending(a, b int) int {
	if a > b {
		return -1
	}
	if a < b {
		return 1
	}
	return 0
}

func compareAbsValue(a, b int) int {
	absA := a
	if a < 0 {
		absA = -a
	}
	absB := b
	if b < 0 {
		absB = -b
	}
	if absA < absB {
		return -1
	}
	if absA > absB {
		return 1
	}
	return 0
}

// Transform function type
type TransformFunc func(int) int

// Transform functions
func square(x int) int {
	return x * x
}

func cube(x int) int {
	return x * x * x
}

func negate(x int) int {
	return -x
}

// StringUtils
type StringUtils struct{}

// String operations
func (su *StringUtils) Reverse(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}

func (su *StringUtils) ToUpper(s string) string {
	return strings.ToUpper(s)
}

func (su *StringUtils) ToLower(s string) string {
	return strings.ToLower(s)
}

func (su *StringUtils) Compare(s1, s2 string) int {
	return strings.Compare(s1, s2)
}

func (su *StringUtils) Concatenate(s1, s2 string) string {
	return s1 + s2
}

func (su *StringUtils) FindChar(s string, ch byte) int {
	for i := 0; i < len(s); i++ {
		if s[i] == ch {
			return i
		}
	}
	return -1
}

// ArrayProcessor with function pointers
type ArrayProcessor struct {
	Data []int
}

// ApplyTransform
func (ap *ArrayProcessor) ApplyTransform(fn TransformFunc) {
	for i := range ap.Data {
		ap.Data[i] = fn(ap.Data[i])
	}
}

// CustomSort
func (ap *ArrayProcessor) CustomSort(cmp CompareFunc) {
	sort.Slice(ap.Data, func(i, j int) bool {
		return cmp(ap.Data[i], ap.Data[j]) < 0
	})
}

// Lambda-like operations with functions
func filterEven(arr []int) []int {
	var result []int
	for _, v := range arr {
		if v%2 == 0 {
			result = append(result, v)
		}
	}
	return result
}

func mapSquare(arr []int) []int {
	result := make([]int, len(arr))
	for i, v := range arr {
		result[i] = v * v
	}
	return result
}

func fold(arr []int, initial int) int {
	result := initial
	for _, v := range arr {
		result += v
	}
	return result
}

// ForEach callback
func forEach(arr []int, callback func(int)) {
	for _, v := range arr {
		callback(v)
	}
}

func main() {
	// Test function pointers with sorting
	arr1 := []int{3, 1, 4, 1, 5, 9, 2, 6}
	arr2 := make([]int, len(arr1))
	copy(arr2, arr1)

	processor1 := &ArrayProcessor{Data: arr1}
	processor1.CustomSort(compareAscending)

	processor2 := &ArrayProcessor{Data: arr2}
	processor2.CustomSort(compareDescending)

	// Test transform
	squareArr := []int{1, 2, 3, 4, 5}
	processor3 := &ArrayProcessor{Data: squareArr}
	processor3.ApplyTransform(square)

	// Test string operations
	su := &StringUtils{}
	str1 := "Hello"
	str2 := "World"

	reversed := su.Reverse(str1)
	upper := su.ToUpper(str1)
	lower := su.ToLower(upper)
	cmpResult := su.Compare(str1, str2)
	concat := su.Concatenate(str1, str2)
	findResult := su.FindChar("Hello World", 'o')

	// Test functional operations
	data := []int{1, 2, 3, 4, 5, 6, 7, 8, 9}
	even := filterEven(data)
	squared := mapSquare(data)
	foldResult := fold(data, 0)

	// Test callbacks
	callbackSum := 0
	forEach(arr1, func(val int) {
		callbackSum += val
	})

	fmt.Printf("Function pointers and strings test: %d %d %d\n",
		len(reversed), findResult, foldResult)
	fmt.Printf("String ops: %s %s\n", upper, concat)
	fmt.Printf("Case: %s, Even count: %d, Squared count: %d\n", lower, len(even), len(squared))
	fmt.Printf("Callback sum: %d, Compare: %d\n", callbackSum, cmpResult)
}

File: benchmark/binary/x86-64/window/small/source/c/function_pointers_and_strings.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

// Function pointer types
typedef int (*compare_func)(const void *, const void *);
typedef int (*transform_func)(int);
typedef void (*callback_func)(int);

// Comparison functions
int compare_int_ascending(const void *a, const void *b) {
    return *(int *)a - *(int *)b;
}

int compare_int_descending(const void *a, const void *b) {
    return *(int *)b - *(int *)a;
}

int compare_abs_value(const void *a, const void *b) {
    int abs_a = abs(*(int *)a);
    int abs_b = abs(*(int *)b);
    return abs_a - abs_b;
}

// Transform functions
int square(int x) {
    return x * x;
}

int cube(int x) {
    return x * x * x;
}

int negate(int x) {
    return -x;
}

int absolute(int x) {
    return x < 0 ? -x : x;
}

// Apply transform function to array
void apply_transform(int *arr, int len, transform_func func) {
    for (int i = 0; i < len; i++) {
        arr[i] = func(arr[i]);
    }
}

// Apply function pointer callback
void for_each(int *arr, int len, callback_func cb) {
    for (int i = 0; i < len; i++) {
        cb(arr[i]);
    }
}

// Callback implementation
void print_value(int val) {
    printf("%d ", val);
}

void accumulate_callback(int val) {
    static int sum = 0;
    sum += val;
}

// Custom sort with function pointer
void custom_sort(int *arr, int len, compare_func cmp) {
    for (int i = 0; i < len - 1; i++) {
        for (int j = 0; j < len - i - 1; j++) {
            if (cmp(&arr[j], &arr[j + 1]) > 0) {
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
            }
        }
    }
}

// String operations
int string_length(const char *str) {
    int len = 0;
    while (*str != '\0') {
        len++;
        str++;
    }
    return len;
}

char* string_reverse(char *str) {
    if (str == NULL) return NULL;
    
    char *start = str;
    char *end = str + string_length(str) - 1;
    
    while (start < end) {
        char temp = *start;
        *start = *end;
        *end = temp;
        start++;
        end--;
    }
    
    return str;
}

char* string_upper(char *str) {
    if (str == NULL) return NULL;
    
    for (int i = 0; str[i] != '\0'; i++) {
        str[i] = toupper(str[i]);
    }
    
    return str;
}

char* string_lower(char *str) {
    if (str == NULL) return NULL;
    
    for (int i = 0; str[i] != '\0'; i++) {
        str[i] = tolower(str[i]);
    }
    
    return str;
}

int string_compare(const char *s1, const char *s2) {
    while (*s1 != '\0' && *s2 != '\0') {
        if (*s1 != *s2) {
            return *s1 - *s2;
        }
        s1++;
        s2++;
    }
    return *s1 - *s2;
}

char* string_concat(char *dest, const char *src, int max_len) {
    int i = 0;
    while (dest[i] != '\0' && i < max_len) {
        i++;
    }
    
    int j = 0;
    while (src[j] != '\0' && i < max_len - 1) {
        dest[i] = src[j];
        i++;
        j++;
    }
    dest[i] = '\0';
    
    return dest;
}

int string_find_char(const char *str, char ch) {
    for (int i = 0; str[i] != '\0'; i++) {
        if (str[i] == ch) {
            return i;
        }
    }
    return -1;
}

// Main function
int main() {
    // Function pointers and transforms
    int arr[] = {3, 1, 4, 1, 5, 9, 2, 6};
    int len = sizeof(arr) / sizeof(arr[0]);
    
    // Test comparisons
    int copy1[8];
    int copy2[8];
    int copy3[8];
    memcpy(copy1, arr, sizeof(arr));
    memcpy(copy2, arr, sizeof(arr));
    memcpy(copy3, arr, sizeof(arr));
    
    custom_sort(copy1, len, compare_int_ascending);
    custom_sort(copy2, len, compare_int_descending);
    custom_sort(copy3, len, compare_abs_value);
    
    // Test transforms
    int copy_square[] = {1, 2, 3, 4, 5};
    apply_transform(copy_square, 5, square);
    
    // String operations
    char str1[] = "Hello";
    char str2[] = "World";
    char result[32] = "Test";
    
    char str_copy[] = "TestString";
    string_reverse(str_copy);
    
    char str_upper[] = "hello";
    string_upper(str_upper);
    
    int cmp_result = string_compare("abc", "abd");
    int find_result = string_find_char("Hello World", 'o');
    
    string_concat(result, " String", sizeof(result));
    
    printf("Operations completed\n");
    
    return 0;
}
